File: app/src/main/java/com/mmjang/ankihelper/ui/plan/PlanEditorActivity.java
package com.mmjang.ankihelper.ui.plan;

import android.content.Intent;
import android.content.pm.PackageManager;
import androidx.annotation.NonNull;
import androidx.core.app.NavUtils;
import android.os.Bundle;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import android.util.Log;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.EditText;
import android.widget.Spinner;
import android.widget.TextView;
import android.widget.Toast;

import com.mmjang.ankihelper.R;
import com.mmjang.ankihelper.anki.AnkiDroidHelper;
import com.mmjang.ankihelper.data.Settings;
import com.mmjang.ankihelper.data.database.AppDatabase;
import com.mmjang.ankihelper.data.plan.OutputPlanEntity;
import com.mmjang.ankihelper.data.plan.OutputPlanPOJO;
import com.mmjang.ankihelper.data.plan.OutputPlanRepository;
import com.mmjang.ankihelper.data.plan.OutputPlanRepositoryHelper;
import com.mmjang.ankihelper.util.Constant;
import com.mmjang.ankihelper.data.dict.DictionaryRegister;
import com.mmjang.ankihelper.data.dict.IDictionary;
import com.mmjang.ankihelper.data.dict.AIDictionary; // Add this import
import com.mmjang.ankihelper.MyApplication;
import com.mmjang.ankihelper.data.plan.OutputPlan;
import com.mmjang.ankihelper.util.Utils;
import com.mmjang.ankihelper.ui.base.BaseEditorActivity;

import org.litepal.crud.LitePalSupport;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

public class PlanEditorActivity extends BaseEditorActivity {

    private String planNameToEdit;
    private AnkiDroidHelper mAnkiDroid;
    private OutputPlanPOJO planForEdit;
    private Map<Long, String> deckList;
    private Map<Long, String> modelList;
    private List<IDictionary> dictionaryList;
    private List<FieldsMapItem> fieldsMapItemList;
    private IDictionary currentDictionary;
    private long currentDeckId;
    private long currentModelId;
    private OutputPlanRepositoryHelper planRepositoryHelper;
    //views
    private EditText planNameEditText;
    private Spinner dictionarySpinner;
    private TextView dictionaryIntroductionTextView;
    private Spinner deckSpinner;
    private Spinner modelSpinner;
    private RecyclerView fieldsSpinnersContainer;

    @Override
    protected void initializeViews() {
        try {
            setViewMember();
            initAnkiApi();

            // Initialize database and load data in background to avoid ANR
            new Thread(new Runnable() {
                @Override
                public void run() {
                    try {
                        // Initialize repository helper off main thread
                        AppDatabase database = AppDatabase.Companion.getInstance(getApplicationContext());
                        OutputPlanRepository repository = new OutputPlanRepository(database.outputPlanDao());
                        planRepositoryHelper = new OutputPlanRepositoryHelper(repository, PlanEditorActivity.this);

                        // Load AnkiDroid data off main thread
                        loadDecksAndModels();

                        // Update UI on main thread
                        runOnUiThread(new Runnable() {
                            @Override
                            public void run() {
                                handleIntent();
                                populateDictionary();
                                populateDecksAndModels();
                            }
                        });
                    } catch (Exception e) {
                        runOnUiThread(new Runnable() {
                            @Override
                            public void run() {
                                Toast.makeText(PlanEditorActivity.this,
                                    "Failed to initialize: " + e.getMessage(),
                                    Toast.LENGTH_SHORT).show();
                            }
                        });
                    }
                }
            }).start();
        } catch (Exception e) {
            Toast.makeText(this, e.getLocalizedMessage(), Toast.LENGTH_SHORT).show();
        }
    }

    @Override
    protected void setupListeners() {
        // Listeners are set in populate methods
    }

    @Override
    protected int getLayoutResId() {
        return R.layout.activity_plan_editor;
    }

    @Override
    protected boolean validateInput() {
        String planName = planNameEditText.getText().toString().trim();
        if (planName.isEmpty()) {
            Toast.makeText(this, R.string.str_plan_name_should_not_be_blank, Toast.LENGTH_SHORT).show();
            return false;
        }
        
        // Check if all fields are empty
        boolean allFieldsAreEmpty = true;
        for (FieldsMapItem item : fieldsMapItemList) {
            String v = item.getExportedElementNames()[item.getSelectedFieldPos()];
            if (!v.equals(Constant.getSharedExportElements()[0])) {
                allFieldsAreEmpty = false;
                break;
            }
        }
        if (allFieldsAreEmpty) {
            Toast.makeText(this, R.string.save_plan_error_all_blank, Toast.LENGTH_SHORT).show();
            return false;
        }
        
        return true;
    }

    @Override
    protected boolean saveData() {
        return savePlan();
    }

    private boolean savePlan() {
        // Check if repository is initialized
        if (planRepositoryHelper == null) {
            Toast.makeText(this, "Please wait, initializing...", Toast.LENGTH_SHORT).show();
            return false;
        }

        // Start async save operation - no blocking!
        savePlanAsync();
        // Return false to prevent immediate finish() - we'll finish after save completes
        return false;
    }

    private void savePlanAsync() {
        final String planName = planNameEditText.getText().toString().trim();

        // Show progress on UI thread
        runOnUiThread(() -> {
            // Disable UI to prevent multiple saves
            planNameEditText.setEnabled(false);
            dictionarySpinner.setEnabled(false);
            deckSpinner.setEnabled(false);
            modelSpinner.setEnabled(false);
        });

        // Step 1: Check for name conflicts (async, non-blocking)
        planRepositoryHelper.getPlanByName(planName, new OutputPlanRepositoryHelper.PlanCallback() {
            @Override
            public void onSuccess(OutputPlanEntity existingPlan) {
                // Check if there's a conflict
                boolean isConflict = false;
                if (planForEdit != null) {
                    // Editing: conflict only if name changed and new name exists
                    if (!planName.equals(planNameToEdit) && existingPlan != null) {
                        isConflict = true;
                    }
                } else {
                    // Creating: conflict if name already exists
                    if (existingPlan != null) {
                        isConflict = true;
                    }
                }

                if (isConflict) {
                    runOnUiThread(() -> {
                        Toast.makeText(PlanEditorActivity.this, R.string.plan_already_exists, Toast.LENGTH_SHORT).show();
                        // Re-enable UI
                        planNameEditText.setEnabled(true);
                        dictionarySpinner.setEnabled(true);
                        deckSpinner.setEnabled(true);
                        modelSpinner.setEnabled(true);
                    });
                } else {
                    // No conflict - proceed with save
                    performSave(planName);
                }
            }

            @Override
            public void onError(Throwable error) {
                // Error checking for conflicts - proceed with save anyway
                performSave(planName);
            }
        });
    }

    private void performSave(final String planName) {
        // Prepare plan data
        OutputPlanPOJO plan;
        if (planForEdit != null) {
            plan = planForEdit;
        } else {
            plan = new OutputPlanPOJO();
        }

        plan.setPlanName(planName);
        plan.setDictionaryKey(currentDictionary.getDictionaryKey());
        plan.setOutputDeckId(currentDeckId);
        plan.setOutputModelId(currentModelId);

        LinkedHashMap<String, String> map = new LinkedHashMap<>();
        for (FieldsMapItem item : fieldsMapItemList) {
            String k = item.getField();
            String v = item.getExportedElementNames()[item.getSelectedFieldPos()];
            map.put(k, v);
        }
        plan.setFieldsMap(map);

        // Convert POJO to Entity
        OutputPlanEntity entity = new OutputPlanEntity();
        entity.setPlanName(plan.getPlanName());
        entity.setDictionaryKey(plan.getDictionaryKey());
        entity.setOutputDeckId(plan.getOutputDeckId());
        entity.setOutputModelId(plan.getOutputModelId());
        entity.setFieldsMap(plan.getFieldsMapString());

        // Perform save operation (async, non-blocking)
        if (planNameToEdit != null) {
            // Update existing plan
            planRepositoryHelper.refreshPlan(planNameToEdit, entity, new OutputPlanRepositoryHelper.OperationCallback() {
                @Override
                public void onSuccess() {
                    runOnUiThread(() -> {
                        Toast.makeText(PlanEditorActivity.this, "Plan saved successfully", Toast.LENGTH_SHORT).show();
                        finish();
                    });
                }

                @Override
                public void onError(Throwable error) {
                    runOnUiThread(() -> {
                        Toast.makeText(PlanEditorActivity.this,
                                "Failed to update plan: " + error.getMessage(),
                                Toast.LENGTH_SHORT).show();
                        // Re-enable UI on error
                        planNameEditText.setEnabled(true);
                        dictionarySpinner.setEnabled(true);
                        deckSpinner.setEnabled(true);
                        modelSpinner.setEnabled(true);
                    });
                }
            });
        } else {
            // Insert new plan
            planRepositoryHelper.savePlan(entity, new OutputPlanRepositoryHelper.OperationCallback() {
                @Override
                public void onSuccess() {
                    runOnUiThread(() -> {
                        Toast.makeText(PlanEditorActivity.this, "Plan saved successfully", Toast.LENGTH_SHORT).show();
                        finish();
                    });
                }

                @Override
                public void onError(Throwable error) {
                    runOnUiThread(() -> {
                        Toast.makeText(PlanEditorActivity.this,
                                "Failed to save plan: " + error.getMessage(),
                                Toast.LENGTH_SHORT).show();
                        // Re-enable UI on error
                        planNameEditText.setEnabled(true);
                        dictionarySpinner.setEnabled(true);
                        deckSpinner.setEnabled(true);
                        modelSpinner.setEnabled(true);
                    });
                }
            });
        }
    }

    private void setViewMember() {
        planNameEditText = (EditText) findViewById(R.id.text_edit_plan_name);
        dictionarySpinner = (Spinner) findViewById(R.id.dictionary_spinner);
        dictionaryIntroductionTextView = (TextView) findViewById(R.id.text_view_dictionary_introduction);
        deckSpinner = (Spinner) findViewById(R.id.deck_spinner);
        modelSpinner = (Spinner) findViewById(R.id.model_spinner);
        fieldsSpinnersContainer = (RecyclerView) findViewById(R.id.recycler_view_fields_map);
    }

    private void initAnkiApi() {
        if (mAnkiDroid == null) {
            mAnkiDroid = new AnkiDroidHelper(this);
        }
        if (!AnkiDroidHelper.isApiAvailable(MyApplication.getContext())) {
            Toast.makeText(this, R.string.api_not_available_message, Toast.LENGTH_LONG).show();
        }

        if (mAnkiDroid.shouldRequestPermission()) {
            mAnkiDroid.requestPermission(this, 0);
        }
    }

    private void handleIntent() {
        Intent intent = getIntent();
        if (intent != null) {
            String action = intent.getAction();
            if (action != null && action.equals(Intent.ACTION_SEND)) {
                String text = intent.getStringExtra(Intent.EXTRA_TEXT);
                if (text != null && !text.isEmpty()) {
                    planNameToEdit = text;
                    // Load plan asynchronously
                    planRepositoryHelper.getPlanByName(planNameToEdit, new OutputPlanRepositoryHelper.PlanCallback() {
                        @Override
                        public void onSuccess(OutputPlanEntity entity) {
                            if (entity != null) {
                                // Convert Entity to POJO
                                planForEdit = new OutputPlanPOJO();
                                planForEdit.setPlanName(entity.getPlanName());
                                planForEdit.setDictionaryKey(entity.getDictionaryKey());
                                planForEdit.setOutputDeckId(entity.getOutputDeckId());
                                planForEdit.setOutputModelId(entity.getOutputModelId());
                                planForEdit.setFieldsMapString(entity.getFieldsMap());
                                //set plan name unable to edit
                                planNameEditText.setText(planNameToEdit);
                                //planNameEditText.setEnabled(false);
                            }
                        }

                        @Override
                        public void onError(Throwable error) {
                            Toast.makeText(PlanEditorActivity.this,
                                    "Failed to load plan: " + error.getMessage(),
                                    Toast.LENGTH_SHORT).show();
                        }
                    });
                }
            }
        }
    }

    private void loadDecksAndModels() {
        deckList = Utils.hashMap2LinkedHashMap(mAnkiDroid.getApi().getDeckList());
        modelList = Utils.hashMap2LinkedHashMap(mAnkiDroid.getApi().getModelList());
    }

    private void populateDictionary() {
        if (dictionaryList == null) {
            dictionaryList = DictionaryRegister.getDictionaryObjectList();
        }

        String[] dictionaryNameList = new String[dictionaryList.size()];
        for (int i = 0; i < dictionaryList.size(); i++) {
            dictionaryNameList[i] = dictionaryList.get(i).getDictionaryName();
        }
        ArrayAdapter<String> dictionarySpinnerAdapter = new ArrayAdapter<>(
                this, android.R.layout.simple_spinner_dropdown_item, dictionaryNameList);
        dictionarySpinner.setAdapter(dictionarySpinnerAdapter);

        if (planForEdit != null) {
            String savedKey = planForEdit.getDictionaryKey();
            boolean found = false;
            for (int i = 0; i < dictionaryList.size(); i++) {
                IDictionary dict = dictionaryList.get(i);
                // Use stable dictionary key for matching
                if (savedKey.equals(dict.getDictionaryKey())) {
                    currentDictionary = dictionaryList.get(i);
                    dictionaryIntroductionTextView.setText(currentDictionary.getIntroduction());
                    dictionarySpinner.setSelection(i);
                    found = true;
                    Log.d("Editor", "Found dictionary: " + dict.getDictionaryName());
                    break;
                }
            }
            if (!found) {
                String message = getString(R.string.error_dictionary_not_found, savedKey);
                Utils.showMessage(PlanEditorActivity.this, message);
            }
        } else {

            int pos = dictionarySpinner.getSelectedItemPosition();
            currentDictionary = dictionaryList.get(pos);
            dictionaryIntroductionTextView.setText(currentDictionary.getIntroduction());

        }

        dictionarySpinner.setOnItemSelectedListener(
                new AdapterView.OnItemSelectedListener() {
                    @Override
                    public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
                        currentDictionary = dictionaryList.get(position);
                        dictionaryIntroductionTextView.setText(currentDictionary.getIntroduction());
                        refreshFieldSpinners();
                    }

                    @Override
                    public void onNothingSelected(AdapterView<?> parent) {

                    }
                }
        );
    }

    private void populateDecksAndModels() {
        ArrayAdapter<String> deckSpinnerAdapter = new ArrayAdapter<>(
                this, android.R.layout.simple_spinner_dropdown_item, Utils.getMapValueArray(deckList));
        deckSpinner.setAdapter(deckSpinnerAdapter);
        ArrayAdapter<String> modelSpinnerAdapter = new ArrayAdapter<>(
                this, android.R.layout.simple_spinner_dropdown_item, Utils.getMapValueArray(modelList));
        modelSpinner.setAdapter(modelSpinnerAdapter);

        if (planForEdit != null) {
            long savedDeckId = planForEdit.getOutputDeckId();
            long savedModelId = planForEdit.getOutputModelId();
            //int i = 0;
            long[] deckIdList = Utils.getMapKeyArray(deckList);
            //int deckPos = Arrays.asList(deckIdList).indexOf(savedDeckId);
            int deckPos = Utils.getArrayIndex(deckIdList, savedDeckId);
            if (deckPos == -1) {
                deckPos = 0;
            }
            currentDeckId = deckIdList[deckPos];
            deckSpinner.setSelection(deckPos);

            long[] modelIdList = Utils.getMapKeyArray(modelList);
            //int modelPos = Arrays.asList(modelIdList).indexOf(savedModelId);
            int modelPos = Utils.getArrayIndex(modelIdList, savedModelId);
            if (modelPos == -1) {
                modelPos = 0;
            }
            currentModelId = modelIdList[modelPos];
            modelSpinner.setSelection(modelPos);

            refreshFieldSpinners();
        } else {
            currentDeckId = Utils.getMapKeyArray(deckList)[0];
            currentModelId = Utils.getMapKeyArray(modelList)[0];
            refreshFieldSpinners();
        }

        modelSpinner.setOnItemSelectedListener(
                new AdapterView.OnItemSelectedListener() {
                    @Override
                    public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
                        currentModelId = Utils.getMapKeyArray(modelList)[position];
                        refreshFieldSpinners();
                    }

                    @Override
                    public void onNothingSelected(AdapterView<?> parent) {

                    }
                }
        );

        deckSpinner.setOnItemSelectedListener(
                new AdapterView.OnItemSelectedListener() {
                    @Override
                    public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
                        currentDeckId = Utils.getMapKeyArray(deckList)[position];
                    }

                    @Override
                    public void onNothingSelected(AdapterView<?> parent) {

                    }
                }
        );

    }

    private void refreshFieldSpinners() {
        String[] fields = mAnkiDroid.getApi().getFieldList(currentModelId);
        String[] dictionaryElements = currentDictionary.getExportElementsList();
        String[] sharedElements = Constant.getSharedExportElements();
        String[] allElements = Utils.concatenate(sharedElements, dictionaryElements);
        fieldsMapItemList = new ArrayList<>();
        //if edit, than set spinner initial position
        if (planForEdit != null && currentModelId == planForEdit.getOutputModelId()) {
            for (String fld : fields) {
                Map<String, String> fldMap = planForEdit.getFieldsMap();
                if (fldMap.containsKey(fld)) {
                    String savedEle = fldMap.get(fld);
                    int pos = Arrays.asList(allElements).indexOf(savedEle);
                    if (pos == -1) {
                        pos = 0;
                    }
                    fieldsMapItemList.add(new FieldsMapItem(fld, allElements, pos));
                }
                //Arrays.asList(allElements).indexOf();
            }
        } else {
            for (String fld : fields) {
                fieldsMapItemList.add(new FieldsMapItem(fld, allElements));
            }
        }

        fieldsSpinnersContainer.setLayoutManager(new LinearLayoutManager(this));
        fieldsSpinnersContainer.setAdapter(new FieldMapSpinnerListAdapter(PlanEditorActivity.this, fieldsMapItemList));
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions,
                                           @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if (requestCode == 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
            //Toast.makeText(TestActivity.this, R.string.permission_granted, Toast.LENGTH_LONG).show();
        } else {
            Toast.makeText(this, R.string.permission_denied, Toast.LENGTH_LONG).show();
        }
    }

}

File: app/src/main/java/com/lmyby/ankihelper/ui/plan/PlanEditorActivity.kt
package com.lmyby.ankihelper.ui.plan

import android.content.Intent
import android.content.pm.PackageManager
import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.AdapterView
import android.widget.ArrayAdapter
import android.widget.EditText
import android.widget.Spinner
import android.widget.TextView
import android.widget.Toast
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.lmyby.ankihelper.MyApplication
import com.lmyby.ankihelper.R
import com.lmyby.ankihelper.anki.AnkiDroidHelper
import com.lmyby.ankihelper.data.database.AppDatabase
import com.lmyby.ankihelper.data.dict.DictionaryRegister
import com.lmyby.ankihelper.data.dict.IDictionary
import com.lmyby.ankihelper.data.plan.OutputPlanEntity
import com.lmyby.ankihelper.data.plan.OutputPlanPOJO
import com.lmyby.ankihelper.data.plan.OutputPlanRepository
import com.lmyby.ankihelper.data.plan.OutputPlanRepositoryHelper
import com.lmyby.ankihelper.ui.base.BaseEditorActivity
import com.lmyby.ankihelper.util.Constant
import com.lmyby.ankihelper.util.Utils
import java.util.LinkedHashMap

/**
 * Activity for creating and editing output plans
 * Converted to Kotlin as part of Phase 5 activity migration
 */
class PlanEditorActivity : BaseEditorActivity() {

    private var planNameToEdit: String? = null
    private lateinit var mAnkiDroid: AnkiDroidHelper
    private var planForEdit: OutputPlanPOJO? = null
    private var deckList: Map<Long, String>? = null
    private var modelList: Map<Long, String>? = null
    private var dictionaryList: List<IDictionary>? = null
    private var fieldsMapItemList: MutableList<FieldsMapItem> = mutableListOf()
    private var currentDictionary: IDictionary? = null
    private var currentDeckId: Long = 0
    private var currentModelId: Long = 0
    private var planRepositoryHelper: OutputPlanRepositoryHelper? = null

    // Views
    private lateinit var planNameEditText: EditText
    private lateinit var dictionarySpinner: Spinner
    private lateinit var dictionaryIntroductionTextView: TextView
    private lateinit var deckSpinner: Spinner
    private lateinit var modelSpinner: Spinner
    private lateinit var fieldsSpinnersContainer: RecyclerView

    override fun getLayoutResId(): Int = R.layout.activity_plan_editor

    override fun initializeViews() {
        try {
            setViewMember()
            initAnkiApi()

            // Initialize database and load data in background to avoid ANR
            Thread {
                try {
                    // Initialize repository helper off main thread
                    val database = AppDatabase.getInstance(applicationContext)
                    val repository = OutputPlanRepository(database.outputPlanDao())
                    planRepositoryHelper = OutputPlanRepositoryHelper(repository, this@PlanEditorActivity)

                    // Load AnkiDroid data off main thread
                    loadDecksAndModels()

                    // Update UI on main thread
                    runOnUiThread {
                        handleIntent()
                        populateDictionary()
                        populateDecksAndModels()
                    }
                } catch (e: Exception) {
                    runOnUiThread {
                        Toast.makeText(
                            this@PlanEditorActivity,
                            "Failed to initialize: ${e.message}",
                            Toast.LENGTH_SHORT
                        ).show()
                    }
                }
            }.start()
        } catch (e: Exception) {
            Toast.makeText(this, e.localizedMessage, Toast.LENGTH_SHORT).show()
        }
    }

    override fun setupListeners() {
        // Listeners are set in populate methods
    }

    override fun validateInput(): Boolean {
        val planName = planNameEditText.text.toString().trim()
        if (planName.isEmpty()) {
            Toast.makeText(this, R.string.str_plan_name_should_not_be_blank, Toast.LENGTH_SHORT).show()
            return false
        }

        // Check if all fields are empty
        val allFieldsAreEmpty = fieldsMapItemList.all { item ->
            val v = item.exportedElementNames[item.selectedFieldPos]
            v == Constant.getSharedExportElements()[0]
        }

        if (allFieldsAreEmpty) {
            Toast.makeText(this, R.string.save_plan_error_all_blank, Toast.LENGTH_SHORT).show()
            return false
        }

        return true
    }

    override fun saveData(): Boolean {
        return savePlan()
    }

    private fun savePlan(): Boolean {
        // Check if repository is initialized
        if (planRepositoryHelper == null) {
            Toast.makeText(this, "Please wait, initializing...", Toast.LENGTH_SHORT).show()
            return false
        }

        // Start async save operation - no blocking!
        savePlanAsync()
        // Return false to prevent immediate finish() - we'll finish after save completes
        return false
    }

    private fun savePlanAsync() {
        val planName = planNameEditText.text.toString().trim()

        // Show progress on UI thread
        runOnUiThread {
            // Disable UI to prevent multiple saves
            planNameEditText.isEnabled = false
            dictionarySpinner.isEnabled = false
            deckSpinner.isEnabled = false
            modelSpinner.isEnabled = false
        }

        // Step 1: Check for name conflicts (async, non-blocking)
        planRepositoryHelper?.getPlanByName(planName, object : OutputPlanRepositoryHelper.PlanCallback {
            override fun onSuccess(existingPlan: OutputPlanEntity?) {
                // Check if there's a conflict
                val isConflict = if (planForEdit != null) {
                    // Editing: conflict only if name changed and new name exists
                    planName != planNameToEdit && existingPlan != null
                } else {
                    // Creating: conflict if name already exists
                    existingPlan != null
                }

                if (isConflict) {
                    runOnUiThread {
                        Toast.makeText(
                            this@PlanEditorActivity,
                            R.string.plan_already_exists,
                            Toast.LENGTH_SHORT
                        ).show()
                        // Re-enable UI
                        planNameEditText.isEnabled = true
                        dictionarySpinner.isEnabled = true
                        deckSpinner.isEnabled = true
                        modelSpinner.isEnabled = true
                    }
                } else {
                    // No conflict - proceed with save
                    performSave(planName)
                }
            }

            override fun onError(error: Throwable) {
                // Error checking for conflicts - proceed with save anyway
                performSave(planName)
            }
        })
    }

    private fun performSave(planName: String) {
        // Prepare plan data
        val plan = planForEdit ?: OutputPlanPOJO()

        plan.planName = planName
        plan.dictionaryKey = currentDictionary?.getDictionaryKey() ?: ""
        plan.outputDeckId = currentDeckId
        plan.outputModelId = currentModelId

        val map = LinkedHashMap<String, String>()
        for (item in fieldsMapItemList) {
            val k = item.field
            val v = item.exportedElementNames[item.selectedFieldPos]
            map[k] = v
        }
        plan.fieldsMap = map

        // Convert POJO to Entity
        val entity = OutputPlanEntity(
            planName = plan.planName,
            dictionaryKey = plan.dictionaryKey,
            outputDeckId = plan.outputDeckId,
            outputModelId = plan.outputModelId,
            fieldsMap = plan.getFieldsMapString()
        )

        // Perform save operation (async, non-blocking)
        val callback = object : OutputPlanRepositoryHelper.OperationCallback {
            override fun onSuccess() {
                runOnUiThread {
                    Toast.makeText(
                        this@PlanEditorActivity,
                        "Plan saved successfully",
                        Toast.LENGTH_SHORT
                    ).show()
                    finish()
                }
            }

            override fun onError(error: Throwable) {
                runOnUiThread {
                    Toast.makeText(
                        this@PlanEditorActivity,
                        "Failed to save plan: ${error.message}",
                        Toast.LENGTH_SHORT
                    ).show()
                    // Re-enable UI on error
                    planNameEditText.isEnabled = true
                    dictionarySpinner.isEnabled = true
                    deckSpinner.isEnabled = true
                    modelSpinner.isEnabled = true
                }
            }
        }

        if (planNameToEdit != null) {
            // Update existing plan
            planRepositoryHelper?.refreshPlan(planNameToEdit!!, entity, callback)
        } else {
            // Insert new plan
            planRepositoryHelper?.savePlan(entity, callback)
        }
    }

    private fun setViewMember() {
        planNameEditText = findViewById(R.id.text_edit_plan_name)
        dictionarySpinner = findViewById(R.id.dictionary_spinner)
        dictionaryIntroductionTextView = findViewById(R.id.text_view_dictionary_introduction)
        deckSpinner = findViewById(R.id.deck_spinner)
        modelSpinner = findViewById(R.id.model_spinner)
        fieldsSpinnersContainer = findViewById(R.id.recycler_view_fields_map)
    }

    private fun initAnkiApi() {
        if (!::mAnkiDroid.isInitialized) {
            mAnkiDroid = AnkiDroidHelper(this)
        }
        if (!AnkiDroidHelper.isApiAvailable(MyApplication.getContext())) {
            Toast.makeText(this, R.string.api_not_available_message, Toast.LENGTH_LONG).show()
        }

        if (mAnkiDroid.shouldRequestPermission()) {
            mAnkiDroid.requestPermission(this, 0)
        }
    }

    private fun handleIntent() {
        intent?.let {
            if (it.action == Intent.ACTION_SEND) {
                val text = it.getStringExtra(Intent.EXTRA_TEXT)
                if (!text.isNullOrEmpty()) {
                    planNameToEdit = text
                    // Load plan asynchronously
                    planRepositoryHelper?.getPlanByName(
                        planNameToEdit!!,
                        object : OutputPlanRepositoryHelper.PlanCallback {
                            override fun onSuccess(entity: OutputPlanEntity?) {
                                entity?.let { plan ->
                                    // Convert Entity to POJO
                                    planForEdit = OutputPlanPOJO().apply {
                                        planName = plan.planName
                                        dictionaryKey = plan.dictionaryKey ?: ""
                                        outputDeckId = plan.outputDeckId
                                        outputModelId = plan.outputModelId
                                        setFieldsMapString(plan.fieldsMap ?: "")
                                    }
                                    // Set plan name unable to edit
                                    planNameEditText.setText(planNameToEdit)
                                    // planNameEditText.isEnabled = false
                                }
                            }

                            override fun onError(error: Throwable) {
                                Toast.makeText(
                                    this@PlanEditorActivity,
                                    "Failed to load plan: ${error.message}",
                                    Toast.LENGTH_SHORT
                                ).show()
                            }
                        }
                    )
                }
            }
        }
    }

    private fun loadDecksAndModels() {
        deckList = Utils.hashMap2LinkedHashMap(mAnkiDroid.api.deckList)
        modelList = Utils.hashMap2LinkedHashMap(mAnkiDroid.api.modelList)
    }

    private fun populateDictionary() {
        if (dictionaryList == null) {
            dictionaryList = DictionaryRegister.getDictionaryObjectList()
        }

        val dictionaryNameList = dictionaryList!!.map { it.getDictionaryName() }.toTypedArray()
        val dictionarySpinnerAdapter = ArrayAdapter(
            this,
            android.R.layout.simple_spinner_dropdown_item,
            dictionaryNameList
        )
        dictionarySpinner.adapter = dictionarySpinnerAdapter

        planForEdit?.let { plan ->
            val savedKey = plan.dictionaryKey
            var found = false
            for (i in dictionaryList!!.indices) {
                val dict = dictionaryList!![i]
                // Use stable dictionary key for matching
                if (savedKey == dict.getDictionaryKey()) {
                    currentDictionary = dictionaryList!![i]
                    dictionaryIntroductionTextView.text = currentDictionary?.getIntroduction()
                    dictionarySpinner.setSelection(i)
                    found = true
                    Log.d("Editor", "Found dictionary: ${dict.getDictionaryName()}")
                    break
                }
            }
            if (!found) {
                val message = getString(R.string.error_dictionary_not_found, savedKey)
                Utils.showMessage(this@PlanEditorActivity, message)
            }
        } ?: run {
            val pos = dictionarySpinner.selectedItemPosition
            currentDictionary = dictionaryList!![pos]
            dictionaryIntroductionTextView.text = currentDictionary?.getIntroduction()
        }

        dictionarySpinner.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
            override fun onItemSelected(parent: AdapterView<*>?, view: View?, position: Int, id: Long) {
                currentDictionary = dictionaryList!![position]
                dictionaryIntroductionTextView.text = currentDictionary?.getIntroduction()
                refreshFieldSpinners()
            }

            override fun onNothingSelected(parent: AdapterView<*>?) {}
        }
    }

    private fun populateDecksAndModels() {
        val deckSpinnerAdapter = ArrayAdapter(
            this,
            android.R.layout.simple_spinner_dropdown_item,
            Utils.getMapValueArray(deckList!!)
        )
        deckSpinner.adapter = deckSpinnerAdapter

        val modelSpinnerAdapter = ArrayAdapter(
            this,
            android.R.layout.simple_spinner_dropdown_item,
            Utils.getMapValueArray(modelList!!)
        )
        modelSpinner.adapter = modelSpinnerAdapter

        planForEdit?.let { plan ->
            val savedDeckId = plan.outputDeckId
            val savedModelId = plan.outputModelId

            val deckIdList = Utils.getMapKeyArray(deckList!!)
            var deckPos = Utils.getArrayIndex(deckIdList, savedDeckId)
            if (deckPos == -1) {
                deckPos = 0
            }
            currentDeckId = deckIdList[deckPos]
            deckSpinner.setSelection(deckPos)

            val modelIdList = Utils.getMapKeyArray(modelList!!)
            var modelPos = Utils.getArrayIndex(modelIdList, savedModelId)
            if (modelPos == -1) {
                modelPos = 0
            }
            currentModelId = modelIdList[modelPos]
            modelSpinner.setSelection(modelPos)

            refreshFieldSpinners()
        } ?: run {
            currentDeckId = Utils.getMapKeyArray(deckList!!)[0]
            currentModelId = Utils.getMapKeyArray(modelList!!)[0]
            refreshFieldSpinners()
        }

        modelSpinner.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
            override fun onItemSelected(parent: AdapterView<*>?, view: View?, position: Int, id: Long) {
                currentModelId = Utils.getMapKeyArray(modelList!!)[position]
                refreshFieldSpinners()
            }

            override fun onNothingSelected(parent: AdapterView<*>?) {}
        }

        deckSpinner.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
            override fun onItemSelected(parent: AdapterView<*>?, view: View?, position: Int, id: Long) {
                currentDeckId = Utils.getMapKeyArray(deckList!!)[position]
            }

            override fun onNothingSelected(parent: AdapterView<*>?) {}
        }
    }

    private fun refreshFieldSpinners() {
        val fields = mAnkiDroid.api.getFieldList(currentModelId)
        val dictionaryElements = currentDictionary?.getExportElementsList() ?: emptyArray()
        val sharedElements = Constant.getSharedExportElements()
        val allElements = Utils.concatenate(sharedElements, dictionaryElements)

        fieldsMapItemList = mutableListOf()

        // If edit, then set spinner initial position
        if (planForEdit != null && currentModelId == planForEdit!!.outputModelId) {
            for (fld in fields) {
                val fldMap = planForEdit!!.fieldsMap
                if (fldMap.containsKey(fld)) {
                    val savedEle = fldMap[fld]
                    var pos = allElements.toList().indexOf(savedEle)
                    if (pos == -1) {
                        pos = 0
                    }
                    fieldsMapItemList.add(FieldsMapItem(fld, allElements, pos))
                }
            }
        } else {
            for (fld in fields) {
                fieldsMapItemList.add(FieldsMapItem(fld, allElements))
            }
        }

        fieldsSpinnersContainer.layoutManager = LinearLayoutManager(this)
        fieldsSpinnersContainer.adapter = FieldMapSpinnerListAdapter(this, fieldsMapItemList)
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<String>,
        grantResults: IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        if (requestCode == 0 && grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
            // Permission granted
        } else {
            Toast.makeText(this, R.string.permission_denied, Toast.LENGTH_LONG).show()
        }
    }
}
